chain-abci/src/enclave_bridge/edp/mod.rs

```diff
@@ -0,0 +1,223 @@
+        Poll::Ready(Ok(()))
+    }
+
+    fn poll_shutdown(self: Pin<&mut Self>, _cx: &mut Context) -> Poll<io::Result<()>> {
+        Poll::Ready(Ok(()))
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct TxValidationApp {
+    inner: TxValidationAsyncStream,
+    rx: Arc<Mutex<Receiver<IntraEnclaveResponse>>>,
+}
+
+impl Default for TxValidationApp {
+    fn default() -> Self {
+        let (tx, rx) = channel();
+        TxValidationApp {
+            inner: TxValidationAsyncStream::new(tx),
+            rx: Arc::new(Mutex::new(rx)),
+        }
+    }
+}
+
+/// It launches a ZMQ server that can server tx-query requests;
+/// (used to be in a separate process -- tx-validation-app that had a custom storage;
+/// now it's in a thread of chain-abci and shares its storage)
+pub fn start_zmq<T: EnclaveProxy + 'static>(
+    proxy: T,
+    zmq_conn_str: &str,
+    network_id: u8,
+    storage: ReadOnlyStorage,
+) -> thread::JoinHandle<()> {
+    let (sender, receiver) = channel();
+    let mut server =
+        server::TxValidationServer::new(zmq_conn_str, proxy, storage, network_id, sender)
+            .expect("could not start a zmq server");
+    log::info!("starting zmq server");
+    let child_t = thread::spawn(move || server.execute());
+    receiver.recv().unwrap();
+    child_t
+}
+
+type UserCallStream = io::Result<Option<Box<dyn AsyncStream>>>;
+
+impl UsercallExtension for TxValidationApp {
+    fn connect_stream<'future>(
+        &'future self,
+        addr: &'future str,
+        _local_addr: Option<&'future mut String>,
+        _peer_addr: Option<&'future mut String>,
+    ) -> Pin<Box<dyn Future<Output = UserCallStream> + 'future>> {
+        async fn connect_stream_inner(
+            stream: TxValidationAsyncStream,
+            addr: &str,
+        ) -> UserCallStream {
+            match addr {
+                "chain-abci" => Ok(Some(Box::new(stream))),
+                _ => Ok(None),
+            }
+        }
+
+        Box::pin(connect_stream_inner(self.inner.clone(), addr))
+    }
+}
+
+impl EnclaveProxy for TxValidationApp {
+    fn check_chain(&mut self, network_id: u8) -> Result<(), ()> {
+        self.process_request(IntraEnclaveRequest::InitChainCheck(network_id))
+            .map(|_| ())
+            .map_err(|_| ())
+    }
+
+    fn process_request(&mut self, request: IntraEnclaveRequest) -> IntraEnclaveResponse {
+        // this lock prevents zmq server + chain-abci to submit requests at the same time
+        // as the zmq is temporary, this can later be removed
+        // (and perhaps use some channel that implements both Sync + Send, unlike mpsc)
+        let rx = self.rx.lock().expect("lock for reply");
+        {
+            let mut stream = self.inner.stream.lock().expect("lock");
+            stream.push_request(request);
+            // release the lock for `UsercallExtension` processing in async streams
+            drop(stream)
```

Comment:
Do we need a manual drop here? Shouldn't the ending of code block automatically drop the stream lock?

Reply:
not necessary, it's more for sanity